addons/binding/org.openhab.binding.rotelra1x/src/main/java/org/openhab/binding/rotelra1x/handler/RotelRa1xHandler.java

```diff
@@ -0,0 +1,345 @@
+                    updateState(CHANNEL_FREQUENCY, readFrequency());
+                } else if ("source".equals(command)) {
+                    updateState(CHANNEL_SOURCE, new StringType(readUntil('!')));
+                } else if ("display".equals(command)) {
+                    String stringLength = readUntil(',');
+                    int length = Integer.parseInt(stringLength);
+                    byte[] data = new byte[length];
+                    int off = 0;
+                    while (off != length) {
+                        int r = serialPort.getInputStream().read(data, off, length - off);
+                        if (r == -1) {
+                            throw new IOException("Connection closed while reading display content");
+                        } else {
+                            off += r;
+                        }
+                    }
+                    // We don't do anything with display content, could add a channel if this is useful
+                } else {
+                    readUntil('!'); // discard
+                }
+
+            } catch (IOException e) {
+                if (serialPort != null) {
+                    logger.debug("Input error while receiving data from amplifier, waiting...", e);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                    disconnect();
+                    try {
+                        Thread.sleep(ERROR_RETRY_DELAY_MS);
+                    } catch (InterruptedException e1) {
+                        return;
+                    }
+                }
+            } catch (Exception e) {
+                if (serialPort != null) { // If serial port is closed, it's set to null,
+                                          // there is no message here,
+                    logger.warn("Unexpected error", e);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE,
+                            "Unknown error while processing input: " + e.getMessage());
+                    disconnect();
+                    try {
+                        Thread.sleep(ERROR_RETRY_DELAY_MS);
+                    } catch (InterruptedException e1) {
+                        return;
+                    }
+                }
+            }
+        }
+    }
+
+    private void send(String text) throws IOException, ConfigurationError {
+        if (power) {
+            connect();
+            serialPort.getOutputStream().write(text.getBytes(StandardCharsets.US_ASCII));
+        }
+    }
+
+    private void sendForce(String text) throws IOException, ConfigurationError {
+        connect();
+        serialPort.getOutputStream().write(text.getBytes(StandardCharsets.US_ASCII));
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        try {
+            if (CHANNEL_POWER.equals(channelUID.getId())) {
+                if (command == OnOffType.ON) {
```

Comment:
I'll keep throwing personal opinions here: I would recommend to create methods for all handling inside the if conditions. Not only for handleVolume an such, but for all. This makes this method much more readable.